### ISSUE TITLE
Field handler is trigger even when resource is first created

### ISSUE THREAD

**Issue by euan-tilley** [nolar/kopf#566]
## Long story short

While #375 appears to have been fixed the field handler is still being triggered when a resource is first created. This didn't happen in 0.26.

## Description

<details><summary>The code snippet to reproduce the issue</summary>

```python
import kopf

@kopf.on.create('company.com', 'v1', 'testresources')
def create_handler(spec, **_):
    pass

@kopf.on.field('company.com', 'v1', 'testresources', field='spec.createField')
def createField_changed(old, new, diff, logger, **_):
    logger.info("Diff - {}".format(diff))
```
</details>

<details><summary>The exact command to reproduce the issue</summary>

```bash
kopf run --standalone test-handler.py

cat <<EOF | kubectl create -f -
apiVersion: company.com/v1
kind: TestResource
metadata:
  name: test
spec:
  createField: true
EOF
```
</details>

<details><summary>The full output of the command that failed</summary>

```
[2020-10-09 18:14:43,256] kopf.activities.auth [INFO    ] Activity 'login_via_client' succeeded.
[2020-10-09 18:14:43,256] kopf.reactor.activit [INFO    ] Initial authentication has finished.
[2020-10-09 18:14:48,484] kopf.objects         [INFO    ] [default/test] Handler 'create_handler' succeeded.
[2020-10-09 18:14:48,600] kopf.objects         [INFO    ] [default/test] Diff - (('add', (), None, True),)
[2020-10-09 18:14:48,601] kopf.objects         [INFO    ] [default/test] Handler 'createField_changed/spec.createField' succeeded.
[2020-10-09 18:14:48,601] kopf.objects         [INFO    ] [default/test] Creation event is processed: 2 succeeded; 0 failed.
```
</details>

## Environment

* Kopf version: 0.28
* Kubernetes version: minikube v0.35.0 on darwin (amd64) & v.1.11.9
* Python version: 3.8.0
* OS/platform: macos & coreos

<details><summary>Python packages installed</summary>

```
aiohttp==3.6.2
aiojobs==0.2.2
async-timeout==3.0.1
attrs==19.3.0
cachetools==4.1.0
certifi==2020.4.5.2
cffi==1.14.0
chardet==3.0.4
click==7.1.2
cryptography==2.9.2
google-auth==1.17.2
idna==2.9
iso8601==0.1.12
Jinja2==2.11.2
kopf==0.28
kubernetes==11.0.0
MarkupSafe==1.1.1
multidict==4.7.6
oauthlib==3.1.0
pip==19.3.1
pyasn1==0.4.8
pyasn1-modules==0.2.8
pycparser==2.20
pycrypto==2.6.1
pykube-ng==20.5.0
pymongo==3.10.1
pyOpenSSL==19.1.0
python-dateutil==2.8.1
PyYAML==5.3.1
requests==2.23.0
requests-oauthlib==1.3.0
rsa==4.6
setuptools==41.6.0
six==1.15.0
typing-extensions==3.7.4.2
urllib3==1.25.9
websocket-client==0.57.0
wheel==0.33.6
yarl==1.4.2
```
</details>

**Comment by nolar** [comment]
Thanks for reporting.

As I understand the issue, the on-field handler is called for the object creation, while you expect it to **not** be called for creation, but only for updates — is this correct?

If so, this is the behaviour which was originally intended for `@on.field` handlers in the first place — react to any event/cause in the context of that field, be that the field update or "field creation" (== a creation of an object with that field defined). 

It worked that way a long time ago but then was broken — by not reacting to creation — up to and including 0.26 (and probably for several versions before that). And was fixed again in #340 (with proper diff checks).

It's my fault that I overlooked that and didn't put it to 0.27 release notes as an incompatible change. And actually didn't notice that this specific scenario was misbehaving and not stablised.


As of now, I can only recommend adding an extra check:


```python
import kopf

@kopf.on.field('zalando.org', 'v1', 'kopfexamples', field='spec.field',
               when=lambda cause, **_: cause.reason != kopf.Reason.CREATE)  # << THIS ONE
def fn(old, new, **_):
    print(f"{old!r} -> {new!r}")

@kopf.on.create('zalando.org', 'v1', 'kopfexamples')
def createfn(**_):
    pass
```

Why `cause.reason != kopf.Reason.CREATE` instead of `old is not None`? For the cases when the field is set for the first time, while the object exists for some time already: it will be a change from `None` to a value, but the reason will be `kopf.Reason.UPDATE`.

---

To give more context on why such a design was intended, I can share some other low- to mid-priority plans (or vision?):

```python
import kopf

@kopf.on.create('kopfexamples', field='spec.field')
def fn(new, **_): pass

@kopf.on.update('kopfexamples', field='status.blah')
def status_changed(old, new, **_): pass

@kopf.on.delete('kopfexamples', field='spec.items')
def should_be_gone(old, **_): pass
```

I.e., putting the creation/update/deletion/resume handlers in the context of specific fields (optionally). Now: they are all applied to the whole object, without diving deeper.

The benefit is at least in having finalizers properly managed for field-handling, but also extending the same semantics to e.g. daemons & timers (though daemons are questionable here):

```python
@kopf.timer('kopfexamples', field='spec.field')
def tick_it(value, **_): pass
```

And more:

```python
@kopf.on.update('kopfexamples', field='status.blah', old=kopf.ABSENT, new=kopf.PRESENT)
def status_added(new, **_): pass

@kopf.on.update('kopfexamples', field='status.blah', new='expected')
def status_changed_as_we_want_it(**_): pass
```

And so on. All in all, the core concept in Kopf's DSL is separation of _declarative_ definitions of the reactions' context (decorators), and _imperative_ definitions of the reactions themselves (handler functions). The fields names and field filters are a part of the "context", not of the "reaction" (normally).

In that case, and in the original intention, the 0.26-like behaviour would be done via `@on.update(..., field=...)`, not via `@on.field(..., field=...)`.

**Neither of this exists now, but waits somewhere on the list to be implemented.**
